### PR TITLE
add DDP HPO support for sigopt. only main_process will have HPO, and …

### PR DESCRIPTION
**Pull request by sywangyi**
…pass argument to other process

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

# What does this PR do?

Fixes # (issue)
HPO do not support DDP now. add support in sigopt backend

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

- trainer: @sgugger

### CODE REVIEW

**Comment by sywangyi** [comment]
@sgugger @yao-matrix. please help to review

File: src/transformers/integrations.py [hunk 1]
```diff
@@ -23,6 +23,8 @@
 import tempfile
 from pathlib import Path
 
+import torch
```
Comment (sgugger) [inline src/transformers/integrations.py:26]: This should be protected behind an `if is_torch_available()` (see failing tests).

File: src/transformers/integrations.py [hunk 2]
```diff
@@ -305,70 +307,95 @@ def dynamic_modules_import_trainable(*args, **kwargs):
 
 
 def run_hp_search_sigopt(trainer, n_trials: int, direction: str, **kwargs) -> BestRun:
+    import pickle
+    from copy import deepcopy
```
Comment (sgugger) [inline src/transformers/integrations.py:311]: Standard library imports should go at the top. Only the one for soft dependencies are delayed.

File: src/transformers/integrations.py [hunk 3]
```diff
@@ -305,70 +307,95 @@ def dynamic_modules_import_trainable(*args, **kwargs):
...
 
-        for run in experiment.loop():
-            with run:
+            while experiment.progress.observation_count < experiment.observation_budget:
+                suggestion = conn.experiments(experiment.id).suggestions().create()
                 trainer.objective = None
-                trainer.train(resume_from_checkpoint=None, trial=run.run)
+                trainer._hp_search_setup(suggestion)
+                if trainer.args.world_size > 1:
+                    torch.distributed.broadcast_object_list(pickle.dumps(trainer.args), src=0)
```
Comment (sgugger) [inline src/transformers/integrations.py:370]: This only works for standard DDP (not on TPU/SageMaker etc). So there should be an early test with a clear error message if we are in a different distributed setup.

File: src/transformers/integrations.py [hunk 4]
```diff
@@ -305,70 +307,95 @@ def dynamic_modules_import_trainable(*args, **kwargs):
...
-            suggestion = conn.experiments(experiment.id).suggestions().create()
+            best = list(conn.experiments(experiment.id).best_assignments().fetch().iterate_pages())[0]
+            best_run = BestRun(best.id, best.value, best.assignments)
+        return best_run
+    else:
+        for i in range(n_trials):
             trainer.objective = None
-            trainer.train(resume_from_checkpoint=None, trial=suggestion)
+            c = list(deepcopy(pickle.dumps(trainer.args)))
+            torch.distributed.broadcast_object_list(c, src=0)
```
Comment (sgugger) [inline src/transformers/integrations.py:389]: Use a better variable name here, please.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR, I left a few comments.
Curious if using `to_dict` and re-instantiating from the dict could have the use of `deepcopy` and pickle here.

**Comment by sywangyi** [comment]
@sgugger thanks for the review. the torch.distributed.broadcast_object_list only support list(each element is pickable) input, does not support dict class

File: src/transformers/integrations.py [hunk 5]
```diff
@@ -308,67 +312,101 @@ def run_hp_search_sigopt(trainer, n_trials: int, direction: str, **kwargs) -> Be
...
+
+            for run in experiment.loop():
+                with run:
+                    trainer.objective = None
+                    trainer._hp_search_setup(run.run)
+                    if trainer.args.world_size > 1:
+                        if trainer.args.parallel_mode != ParallelMode.DISTRIBUTED:
+                            raise RuntimeError("only support DDP Sigopt HPO for ParallelMode.DISTRIBUTED currently.")
+                        if not is_torch_available():
+                            raise RuntimeError("only support DDP Sigopt HPO using torch DDP.")
```
Comment (sgugger) [inline src/transformers/integrations.py:338]:
```suggestion
```
The `Trainer` requires torch already, so this check is not needed :-)

**Review by sgugger — APPROVED** [review]
Thanks for iterating!
Can you just solve the merge conflicts and address the small comment below? Then we should be good to merge.

**Pull request merged**